scrapy/settings/__init__.py

```diff
@@ -34,9 +40,20 @@ def __init__(self, value, priority):
 
     def set(self, value, priority):
         """Sets value if priority is higher or equal than current priority."""
-        if priority >= self.priority:
-            self.value = value
-            self.priority = priority
+        if isinstance(self.value, BaseSettings) and isinstance(value, Mapping):
+            # Ignore self.priority if self.value has per-key priorities
+            # XXX: Second condition allows completely replacing a BaseSettings
+            #      instance with a value of incompatible type (i.e. something
+            #      that does not map keys to values, e.g.a list). I included
+            #      this b/c the test_cmdline/__init__.py test uses a deprecated
+            #      configuration API (EXTENSIONS is a list) and fails otherwise.
+            #      Should it stay in?
+            self.value.update(value, priority)
+            self.priority = max((priority, self.priority))
```

Comment:
Idea here is that `self.priority` is the highest priority within the internal `BaseSettings`? In that case, I think we should lookup inside the provided `value` instead of relying in the passed `priority`. Examples like this one won't get the expected priority otherwise:

```python
> s = Settings()
> s.set('option', BaseSettings({'key': 'value'}, priority='spider'))
> s.attributes['option'].priority
  20  # priority 'project' instead of 'spider'
```